データベース設計

eigo-kajihara opened the pull request:
# what
データベース設計

# why
必要な機能確認のため

eigo-kajihara committed:
first_commit

YukiYanagikubo commented on README.md:
```diff
@@ -1,24 +1,43 @@
+- has_many :groups through: :members
+
+## messagesテーブル
+|Column |Type   |Options          |
```
どのグループのメッセージなのかを区別するgroup_id必要になります！

eigo-kajihara replied:
確認しました。

YukiYanagikubo commented on README.md:
```diff
@@ -1,24 +1,43 @@
+|--------|-------|-----------------|
+|name    |integer|foreign_key: true|
+
+### Association
```
groupのアソシエーションも見直してみましょう！
中間テーブルを含んだ場合のアソシエーションを調べてみましょう！

eigo-kajihara replied:
確認しました。

ga-ladder replied:
through: :paramsが付け加えられているのは良いと思います！
ただ、これだけでは機能しません。

- has_many :members

を中間テーブルをhas_many:users through::members の上に記述しておきましょう

ga-ladder replied:
また、messagesテーブルは記述しているようなので忘れているだけかと思いますが、チャットグループにはそれぞれの部屋毎にメッセージが管理できるようになっているはずなので、そちらのアソシエーションも設定するようにしてください。具体的には、groupsテーブルとmessagesテーブルを関連づける設定です。

YukiYanagikubo commented on README.md:
```diff
@@ -1,24 +1,43 @@
-* Deployment instructions
-
-* ...
+## usersテーブル
```
オプションの付け方についても調べてみましょう！

eigo-kajihara replied:
確認しました。

eigo-kajihara committed:
update_readme

編集点
Usersテーブルのオプション
Messageテーブルのオプション
Membersテーブルのオプション
Groupsテーブルのタイプとオプション
Groupsテーブルのアソシエーション

ga-ladder commented on README.md:
```diff
@@ -1,43 +1,44 @@
+|Column            |Type    |Options    |
+|------------------|--------|-----------|
+|encrypted_password|string  |null: false|
+|name              |string  |null: false|
```
chat space では名前で検索する機能を実装するので、 :name に add_index を追加しておきましょう
他のColumnでadd_indexが付いているところがありますが、一般的にindexは特定のカラムを検索して取り出すときに使われますのでもう一度必要かどうか考えてみてください。

ga-ladder replied:
:name は 重複を避けるためにもunique: true のオプションを付けておくのがよいと思います。

eigo-kajihara replied:
確認しました。

ga-ladder commented on README.md:
```diff
@@ -1,43 +1,44 @@
+|email             |string  |null: false|
 
 ### Association
 - has_many :messages
```
下にも書いていますが、中間テーブルとのアソシエーションを設定するときには、
- has_many :members
も必要になります。

eigo-kajihara replied:
確認しました。

eigo-kajihara committed:
Update_readme

編集点
Nameのオプション
Usersテーブルのアソシエーション
User_idのオプション
group_idのオプション

eigo-kajihara committed:
Revert "Update_readme"

This reverts commit 716ebcf155f6ce1c72113b7ae7b728e870c9d1ac.

eigo-kajihara committed:
update_readme

編集点
Nameのオプション
Usersテーブルのアソシエーション
Groupsテーブルのアソシエーション

hzak2xx commented on README.md:
```diff
@@ -1,24 +1,46 @@
+|------------------|--------|------------------------------------|
+|encrypted_password|string  |null: false                         |
+|name              |string  |null: false, unique: true, add_index|
+|email             |string  |null: false                         |
```
email, passwordはdeivseで生成されるので書く必要はないですね

hzak2xx commented on README.md:
```diff
@@ -1,24 +1,46 @@
+|Column  |Type   |Options          |
+|--------|-------|-----------------|
+|image   |text   |null: false      |
+|text    |text   |null: false      |
```
text, imageはどちからあればいいのでnotnull制約は必要ないです

hzak2xx commented on README.md:
```diff
@@ -1,24 +1,46 @@
+|image   |text   |null: false      |
+|text    |text   |null: false      |
+|user_id |integer|foreign_key: true|
+|group_id|integer|foreign_key: true|
```
null: false, indexをつけたいです

hzak2xx commented on README.md:
```diff
@@ -1,24 +1,46 @@
+
+|Column  |Type  |Options                             |
+|--------|------|------------------------------------|
+|name    |string|null: false, unique: true, add_index|
```
ここはindex必要ないです

eigo-kajihara commented:
グループへの招待時にネーム検索からメンバーを招待しているので、nameにindexを貼ったのですが、何故nameにindexは必要ないのでしょうか？

ga-ladder commented:
> グループへの招待時にネーム検索からメンバーを招待しているので、nameにindexを貼ったのですが、何故nameにindexは必要ないのでしょうか？

groupsテーブルにおけるnemeカラムはグループ名のことですので、今回のchat spaceにおいては検索では使われなさそうですね。
eigo-kajiharaさんがおっしゃっているのは、usersテーブルのnameカラムのことではないでしょうか？

eigo-kajihara committed:
update_readme

編集
usersテーブルのオプション
messagesテーブルのオプション
membersテーブルのオプション
groupsテーブルのオプション

eigo-kajihara commented:
理解しました。

ga-ladder commented on README.md:
```diff
@@ -1,24 +1,44 @@
+|name    |string|null: false, unique: true|
+
+### Association
+_ has_many :messages
```
データベースとは関係ないですが、
アンダーバーになってます

ga-ladder commented on README.md:
```diff
@@ -1,24 +1,44 @@
+|------------------|--------|------------------------------------|
+|name              |string  |null: false, unique: true, add_index|
+
+### Association
```
中間テーブル(messagesテーブル)を利用したアソシエーションはgroupsテーブルで記述されているように、usersテーブルとmessagesテーブルとのアソシエーションも記述する必要があります。

ga-ladder commented on README.md:
```diff
@@ -1,24 +1,44 @@
+## messagesテーブル
+|Column  |Type   |Options                                  |
+|--------|-------|-----------------------------------------|
+|image   |text   |null: false                              |
```
ここをnull: falseにしてしまうと、毎回のメッセージに画像を挿入する必要が出てきます。

eigo-kajihara committed:
update_readme

編集
usersテーブルのアソシエーション
messagesテーブルのオプション、アソシエーション

ga-ladder commented on README.md:
```diff
@@ -1,24 +1,45 @@
+### Association
+- has_many :messages
+- has_many :members
+- has_many :groups through: :members
```
has_many :groups とthrough: :members の間には , が必要です。
これはgroupsテーブルのところでも同様です。

eigo-kajihara committed:
update_readme

編集
Usersテーブルのアソシエーション
Messagesテーブルのアソシエーション

ga-ladder commented:
LGTMです！

eigo-kajihara commented:
添削有り難う御座いました。
